src/util/expr_iterator.h

```diff
@@ -0,0 +1,242 @@
+      {
+        break;
+      }
+      m_stack.back().it++;
+    }
+    return this->downcast();
+  }
+
+  /// Dereference operator
+  /// Dereferencing end() iterator is undefined behaviour
+  const exprt &operator*() const
+  {
+    PRECONDITION(!m_stack.empty());
+    return m_stack.back().expr.get();
+  }
+
+  /// Dereference operator (member access)
+  /// Dereferencing end() iterator is undefined behaviour
+  const exprt *operator->() const
+  {
+    return &**this;
+  }
+
+protected:
+  /// Create end iterator
+  depth_iterator_baset() { }
+
+  /// Create begin iterator, starting at the supplied node
+  explicit depth_iterator_baset(const exprt &root)
+  {
+    this->downcast().push_expr(root);
+  }
+
+  /// Obtain non-const exprt reference. Performs a copy-on-write on
+  /// the root node as a side effect. To be called only if a the root
+  /// is a non-const exprt. Do not call on the end() iterator
+  exprt &mutate()
+  {
+    PRECONDITION(!m_stack.empty());
+    auto expr=std::ref(const_cast<exprt &>(m_stack.front().expr.get()));
+    for(auto &state : m_stack)
+    {
+      auto &operands=expr.get().operands();
+      const auto i=operands.size()-(state.end-state.it);
+      const auto it=operands.begin()+i;
+      state.expr=expr.get();
+      state.it=it;
+      state.end=operands.end();
+      if(!(state==m_stack.back()))
+      {
+        expr=std::ref(*it);
+      }
+    }
+    return expr.get();
+  }
+
+  /// Pushes expression onto the stack
+  /// If overriden, this function should be called from the inheriting
+  /// class by the override function
+  /// \return true if element was successfully pushed onto the stack,
+  /// false otherwise
+  /// If returning false, child will not be iterated over
+  bool push_expr(const exprt &expr)
+  {
+    m_stack.emplace_back(expr, expr.operands().begin(), expr.operands().end());
+    return true;
+  }
+
+private:
+  typedef typename std::vector<exprt>::const_iterator operands_iteratort;
+
+  struct expr_statet final
+  {
+    expr_statet(
+      const exprt &expr,
+      operands_iteratort it,
+      operands_iteratort end):
+      expr(expr), it(it), end(end) { }
+    expr_statet(const expr_statet &other):
+      expr(other.expr), it(other.it), end(other.end) { }
+    std::reference_wrapper<const exprt> expr;
+    operands_iteratort it;
+    operands_iteratort end;
+    bool operator==(const expr_statet &other) const
+    {
+      return it==other.it && expr.get()==other.expr.get();
+    }
+  };
+
+  std::vector<expr_statet> m_stack;
+
+  depth_iterator_t &downcast()
+  {
+    return static_cast<depth_iterator_t &>(*this);
+  }
+};
+
+template<typename depth_iterator_t>
+class unique_depth_iterator_baset:
+  public depth_iterator_baset<depth_iterator_t>
+{
+  friend depth_iterator_baset<depth_iterator_t>;
+protected:
+  explicit unique_depth_iterator_baset(const exprt &expr):
+    depth_iterator_baset<depth_iterator_t>()
+  { this->push_expr(expr); }
```

Comment:
> It is a bad practice, but virtual functions do not have advantage here.

I'm not suggesting they do. I'm pointing out that the similarity of CRTP to dynamic dispatch suggests that the same guidelines should be observed i.e. don't call pseudo-virtual methods from constructors.